### PR TITLE
Feature/mob 4166 refine analytics payment - APPROACH A

### PR DESCRIPTION
**Pull request by BartlomiejSopalaKarhoo**
## JIRA
[MOB-4166](https://karhoo.atlassian.net/browse/MOB-4166)

## DESCRIPTION

**APPROACH A** 
This pr presents how uisdk / client app will change if we decide to extend current `Result` structure and create tuples in success/failure case, and add correlationId as a second component.

Related [Network SDK pull request](https://github.com/karhoo/karhoo-ios-sdk/pull/153)
## Types of changes

What types of changes does your code introduce to the project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix 
- [ ] New feature 
- [ ] Tech Debt
- [ ] Breaking change 
- [ ] Documentation Update
- [ ] Other: *define what type of change it is*

## CHECKLIST

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. This is simply a reminder of what we are going to look for before merging code._
Please review the guidelines for contributing to this repository.

- [ ] All work relating to the ticket is complete (Code complete/Acceptance Criteria met)
- [ ] Lint and unit tests pass locally (if appropriate)
- [ ] Added tests that prove the fix is effective or that feature works (if appropriate)
- [ ] Documentation (if appropriate)

## SCREENSHOTS
[Add screenshots if any for changes you have made to the App/UI]

### CODE REVIEW

File: KarhooUISDK/Workers/Analytics.swift [hunk 1]
```diff
@@ -18,9 +18,14 @@ public protocol Analytics {
     func pickupAddressSelected(locationDetails: LocationInfo)
     func destinationAddressSelected(locationDetails: LocationInfo)
     
-    func bookingRequested(tripDetails: TripInfo)
-    func paymentSucceed()
-    func paymentFailed(message: String, last4Digits: String, date: Date, amount: String, currency: String)
+    func bookingRequested(tripDetails: TripInfo, outboundTripId: String?)
+    func paymentSucceed(tripDetails: TripInfo)
```
Comment (dianakarh) [inline KarhooUISDK/Workers/Analytics.swift:22]: names need updating to the new standards (success, failure, requested)

File: KarhooUISDK/Screens/CheckoutScreen/Checkout/KarhooCheckoutPresenter.swift [hunk 2]
```diff
@@ -526,38 +524,59 @@ final class KarhooCheckoutPresenter: CheckoutPresenter {
...
 
-    private func reportPaymentFailure(_ message: String) {
-        analytics.paymentFailed(
-                message: message,
-                last4Digits: retrievePaymentNonce()?.lastFour ?? "",
-                date: Date(),
-                amount: quote.price.highPrice.description,
-                currency: quote.price.currencyCode
+    private func reportBookingSuccess(tripId: String, quoteId: String, correlationId: String?) {
+            analytics.bookingSucceed(tripId: tripId, quoteId: quoteId, correlationId: correlationId)
```
Comment (AleksanderWedrychowskiKarhoo) [inline KarhooUISDK/Screens/CheckoutScreen/Checkout/KarhooCheckoutPresenter.swift:564]: I think you should apply auto-formating (control+I in Xcode) to the file.

File: KarhooUISDK/UIComponents/LoyaltyView/KarhooLoyaltyPresenter.swift [hunk 3]
```diff
@@ -425,7 +438,13 @@ final class KarhooLoyaltyPresenter: LoyaltyPresenter {
             flexpay: flexPay, membership: ""
         )
         
-        loyaltyService.getLoyaltyPreAuth(preAuthRequest: request).execute { result in
+        loyaltyService.getLoyaltyPreAuth(preAuthRequest: request).execute { [weak self] result in
+            if let _ = result.successValue(), let currentMode = self?.currentMode {
```
Comment (AleksanderWedrychowskiKarhoo) [inline KarhooUISDK/UIComponents/LoyaltyView/KarhooLoyaltyPresenter.swift:442]: `result.successValue() != nil` preferred over `let _ = result.successValue()`

File: KarhooUISDK/UIComponents/LoyaltyView/KarhooLoyaltyPresenter.swift [hunk 4]
```diff
@@ -443,4 +462,34 @@ final class KarhooLoyaltyPresenter: LoyaltyPresenter {
...
+    
+    
+    private func reportLoyaltyStatusRequested(
+        quoteId: String?,
+        correlationId: String?,
+        loyaltyName: String?,
+        loyaltyStatus: LoyaltyStatus?,
+        errorSlug: String?,
+        errorMessage: String?
+
```
Comment (AleksanderWedrychowskiKarhoo) [inline KarhooUISDK/UIComponents/LoyaltyView/KarhooLoyaltyPresenter.swift:484]: Empty line?

File: KarhooUISDK/Workers/Analytics.swift [hunk 5]
```diff
@@ -223,6 +372,20 @@ func bookingScreenOpened() {
...
+        static let loyaltyStatusCanBurn = "loyalty_status_can_burn"
+        static let loyaltyStatusCanEarn = "loyalty_status_can_earn"
+        static let loyaltyStatusBalance = "loyalty_status_balance"
+        static let errorSlug = "error_slug"
+        static let errorMessage = "error_message"
+        static let loyaltyName = "loyaltyName"
+        static let loyaltyPreauthType = "loyalty_preauth_type"
+
+        
+
```
Comment (AleksanderWedrychowskiKarhoo) [inline KarhooUISDK/Workers/Analytics.swift:388]: Empty lines. SwiftLint isn't showing the warning?

**Review by AleksanderWedrychowskiKarhoo — CHANGES_REQUESTED** [review]

**Review by AleksanderWedrychowskiKarhoo — APPROVED** [review]

**Review by AleksanderWedrychowskiKarhoo — DISMISSED** [review]

**Review by AleksanderWedrychowskiKarhoo — APPROVED** [review]

File: KarhooUISDK/UIComponents/LoyaltyView/KarhooLoyaltyPresenter.swift [hunk 6]
```diff
@@ -424,23 +430,66 @@ final class KarhooLoyaltyPresenter: LoyaltyPresenter {
...
+        quoteId: String?,
+        correlationId: String?,
+        loyaltyName: String?,
+        loyaltyStatus: LoyaltyStatus?,
+        error: KarhooError?
+    ) {
+        analytics.loyaltyStatusRequested(
+            quoteId: quoteId,
+            correlationId: correlationId,
+            loyaltyName: correlationId,
```
Comment (dianakarh) [inline KarhooUISDK/UIComponents/LoyaltyView/KarhooLoyaltyPresenter.swift:489]: @BartlomiejSopalaKarhoo correlation id passed as loyalty name

File: KarhooUISDK/Workers/Analytics.swift [hunk 7]
```diff
@@ -17,10 +17,14 @@ public protocol Analytics {
     func userCalledDriver(trip: TripInfo?)
     func pickupAddressSelected(locationDetails: LocationInfo)
     func destinationAddressSelected(locationDetails: LocationInfo)
-    
-    func bookingRequested(tripDetails: TripInfo)
-    func paymentSucceed()
-    func paymentFailed(message: String, last4Digits: String, date: Date, amount: String, currency: String)
+    func bookingRequested(quoteId: String)
```
Comment (dianakarh) [inline KarhooUISDK/Workers/Analytics.swift:20]: the fleets shown event is never called. is it in another PR?

Comment (dianakarh) [reply KarhooUISDK/Workers/Analytics.swift:20]: found it. it's already merged

**Review by dianakarh — APPROVED** [review]

**Review by AleksanderWedrychowskiKarhoo — APPROVED** [review]

**Review by dianakarh — APPROVED** [review]

**Pull request merged**